### PR TITLE
Std futures redux

### PR DESCRIPTION
**Pull request by Ralith**
- [x] tests/examples
- [x] interop tool
- [x] h3

### CODE REVIEW

**Comment by stammw** [comment]
CI failure seems to happen because it is trying to build on `stable`?

**Comment by Ralith** [comment]
I can't find a way to make our CI use nightly without redoing it entirely :(
https://github.com/crate-ci/azure-pipelines/issues/58

**Comment by Ralith** [comment]
Tidied up some regressions from the rebase.

**Comment by Ralith** [comment]
Reapplied tidying. You've got two partial revert commits that keep coming back, @stammw; just rebase -i and drop them if that happens again.

**Comment by Ralith** [comment]
Also fixed a deleted `use` that crept in.

File: quinn-h3/src/client.rs [hunk 1]
```diff
@@ -48,21 +45,14 @@ impl Builder {
...
-            endpoint_driver,
-            Client {
-                endpoint,
-                settings: self.settings,
-                log: self
-                    .log
-                    .unwrap_or_else(|| Logger::root(slog::Discard, o!())),
-            },
-        ))
+    pub fn endpoint(self, endpoint: Endpoint) -> Client {
```
Comment (Ralith) [inline quinn-h3/src/client.rs]: Nice change! I like that we can now share endpoints between multiple types of outgoing connections.

**Comment by Ralith** [comment]
CI's passing, this is ready for final review!

File: interop/Cargo.toml [hunk 2]
```diff
@@ -3,27 +3,27 @@ name = "interop"
...
+#quinn-h3 = { path = "../quinn-h3" }
 quinn-proto = { path = "../quinn-proto" }
 slog-term = "2"
 bytes = "0.4.7"
 structopt = "0.3.0"
-tokio = "0.1.6"
-tokio-current-thread = "0.1"
-tokio-reactor = "0.1"
+tokio = "0.2.0-alpha.5"
+tokio-net = "0.2.0-alpha.5"
```
[the remark this answers is not in the document]
Comment (Ralith) [reply interop/Cargo.toml]: We need this for `tokio_net::driver::Handle`, which `tokio` does not reexport.

[the remark this answers is not in the document]
Comment (djc) [reply interop/Cargo.toml]: Ah. Little comment in `Cargo.toml`?

File: quinn/src/streams.rs [hunk 3]
```diff
@@ -183,26 +186,64 @@ impl Drop for SendStream {
...
+            this.is_0rtt,
+            &mut this.finishing,
+            cx,
+        )
+    }
+}
 
-    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        self.stream.poll_finish()
+fn poll_finish(
```
[the remark this answers is not in the document]
Comment (Ralith) [reply quinn/src/streams.rs]: I had difficulty factoring out the common state in `SendStream` and `Finish`; I'll have another go tonight, though, it really does seem like it should be possible to do better.

[the remark this answers is not in the document]
Comment (Ralith) [reply quinn/src/streams.rs]: I modified `Finish` to borrow `self` and everything was much saner! Finishing a stream concurrently with other operations on it doesn't make sense, so this was probably for the best.

File: interop/src/main.rs [hunk 4]
```diff
@@ -93,228 +104,249 @@ fn run(log: Logger, options: Opt) -> Result<()> {
...
-                        conn.open_bi()
-                            .map_err(|e| format_err!("failed to open 0-RTT stream: {}", e))
-                    })
-                    .and_then(|(send, recv)| get(send, recv))
-                    .map(|_| {
-                        zero_rtt = true;
+    runtime.spawn(endpoint_driver.unwrap_or_else(|e| eprintln!("IO error: {}", e)));
+
+    runtime.spawn({
+        let endpoint = endpoint.clone();
```
[the remark this answers is not in the document]
Comment (Ralith) [reply interop/src/main.rs]: That's way better, I should have done this ages ago!

**Pull request merged**